File: examples/typescript/src/presence.ts
// Presence Example
// -----------------
// Demonstrates: setting presence, tracking peers, and reacting to presence changes
//
// Run: npm run presence
// Requires: Starfish server running at ws://localhost:8080/starfish
//
// This example creates two clients. Each sets presence data, and both
// observe each other's presence updates.

import { StarfishClient } from "@starfish/client";
import WebSocket from "ws";

const SERVER_URL = "ws://localhost:8080/starfish";

function createClient(name: string) {
  return new StarfishClient({
    server: SERVER_URL,
    ws: (url) => new WebSocket(url) as any,
    client: { name },
  });
}

async function main() {
  const alice = createClient("Alice");
  const bob = createClient("Bob");

  await alice.connect();
  await bob.connect();
  await alice.join("presence-demo");
  await bob.join("presence-demo");
  console.log("Alice and Bob joined the session.");

  // Watch presence changes from Alice's perspective
  alice.presence$.subscribe((presenceMap) => {
    if (presenceMap.size === 0) return;
    console.log("\nAlice sees presence:");
    for (const [peerId, data] of presenceMap) {
      console.log(`  ${peerId}: ${JSON.stringify(data)}`);
    }
  });

  // Watch the peer list from Bob's perspective
  bob.peers$.subscribe((peers) => {
    console.log(`Bob sees ${peers.length} peer(s): ${peers.map((p) => p.name).join(", ")}`);
  });

  await sleep(200);

  // Set presence for each client
  console.log("\nAlice sets presence: { status: 'active', color: 'blue' }");
  alice.presence.set({ status: "active", color: "blue" });
  await sleep(300);

  console.log("Bob sets presence: { status: 'away', color: 'red' }");
  bob.presence.set({ status: "away", color: "red" });
  await sleep(300);

  // Update presence
  console.log("Alice updates presence: { status: 'typing', color: 'blue' }");
  alice.presence.set({ status: "typing", color: "blue" });
  await sleep(300);

  // Clean up
  await alice.disconnect();
  await bob.disconnect();
  console.log("\nDone.");
}

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: examples/python/presence.py
# Presence Example
# -----------------
# Demonstrates: setting presence, tracking peers, and reacting to presence changes
#
# Run: python presence.py
# Requires: Starfish server running at ws://localhost:8080/starfish
#
# This example creates two clients. Each sets presence data, and both
# observe each other's presence updates.

import asyncio
import json
import os

from starfish import StarfishClient, StarfishClientOptions, ClientIdentity

SERVER_URL = os.environ.get("STARFISH_SERVER_URL", "ws://localhost:8080/starfish")


def create_client(name: str) -> StarfishClient:
    return StarfishClient(
        StarfishClientOptions(server=SERVER_URL, client=ClientIdentity(name=name))
    )


async def main():
    alice = create_client("Alice")
    bob = create_client("Bob")

    await alice.connect()
    await bob.connect()
    await alice.join("presence-demo")
    await bob.join("presence-demo")
    print("Alice and Bob joined the session.")

    # Watch presence changes from Alice's perspective
    def on_presence(presence_map):
        if not presence_map:
            return
        print("\nAlice sees presence:")
        for peer_id, data in presence_map.items():
            print(f"  {peer_id}: {json.dumps(data)}")

    alice.presence.subscribe(on_presence)

    # Watch the peer list from Bob's perspective
    bob.peers.subscribe(
        lambda peers: print(f"Bob sees {len(peers)} peer(s): {', '.join(p.name or '' for p in peers)}")
    )

    await asyncio.sleep(0.2)

    # Set presence for each client
    print("\nAlice sets presence: { status: 'active', color: 'blue' }")
    alice.presence_set({"status": "active", "color": "blue"})
    await asyncio.sleep(0.3)

    print("Bob sets presence: { status: 'away', color: 'red' }")
    bob.presence_set({"status": "away", "color": "red"})
    await asyncio.sleep(0.3)

    # Update presence
    print("Alice updates presence: { status: 'typing', color: 'blue' }")
    alice.presence_set({"status": "typing", "color": "blue"})
    await asyncio.sleep(0.3)

    # Clean up
    await alice.disconnect()
    await bob.disconnect()
    print("\nDone.")


if __name__ == "__main__":
    asyncio.run(main())
